src/Text/Pandoc/Writers/LaTeX.hs

```diff
@@ -1472,18 +1480,36 @@ citationsToBiblatex (c:cs)
                     NormalCitation -> "\\autocite"
       return $ text cmd <>
                braces (literal (T.intercalate "," (map citationId (c:cs))))
-  | otherwise = do
-    let cmd = case citationMode c of
+  | otherwise 
+    = do  
+      let cmd = case citationMode c of
                     SuppressAuthor -> "\\autocites*"
                     AuthorInText   -> "\\textcites"
                     NormalCitation -> "\\autocites"
-    let convertOne Citation { citationId = k
-                            , citationPrefix = p
-                            , citationSuffix = s
-                            }
-                = citeArguments p s k
-    args <- mapM convertOne (c:cs)
-    return $ text cmd <> foldl' (<>) empty args
+
+      let groups = map correctOrder $ reverse $ foldl' grouper [([[]], [])] (c:cs)
+      docGroups <- mapM citeArgsList groups      
+
+      return $ text cmd <> (mconcat docGroups)
+
+  where grouper prev cit = grouper' (citationPrefix cit)
+               (citationSuffix cit) (citationId cit) 
+            where grouper' [] sfx cid 
+                     | null sfx = addToGroup id 
+                     | otherwise = addToGroup ((:) sfx) 
+                     where addToGroup fn = case prev of  
+                             (((pfx:pfxs), ids):rest) 
+                                 | null pfxs -> (fn [pfx], cid:ids):rest
+                                 | otherwise -> (fn $ pfxs, [cid]):prev
+                             (([], ids):rest) -> (fn [[]], cid:ids):rest
+                             [] -> (fn [[]], [cid]):prev
+                  grouper' pfx [] cid = ([pfx], [cid]):prev
+                  grouper' pfx sfx cid = ([sfx, pfx], [cid]):prev 
+
+        correctOrder (pfxs, ids) = (reverse pfxs, reverse ids)
+        citeArgsList ((p:s:_), k) = citeArgumentsList p s k
+        citeArgsList ((p:_), k) = citeArgumentsList p [] k
+        citeArgsList ([], k) = citeArgumentsList [] [] k
```

Comment:
This is getting a lot clearer but I still think it could be made more perspicuous.

As I understand it, `foldl' grouper` produces a list of pairs `(affixes, citeIds)` in which `affixes` is a list with either 0, 1, or 2 members.  If 0, we have neither prefix nor suffix. If 1, it's a prefix.  If 2, then the first is a prefix and the second is a suffix.  The function `citeArgsList` takes this list and converts it into two values, for prefix and suffix, which it passes to `citeArgumentsList`.  This all seems overly complicated and a bit fragile.  We can encode this better in the types.  Why not have the fold contsruct a list of triples `(prefix, suffix, ids)` and then change the type of `citeArgumentList` so that it can *directly* consume these triples?  THen you avoid the need for `citeArgsList` and simplify the code.  Or, even more haskelly, you could introduce a data type
```
type Prefix = [Inline]
type Suffix = [Inline]
type CiteId = Text
data CiteGroup = Prefix Suffix [CiteId]
```

I think this change would also allow you to simplify the definition of `grouper` and make it a lot clearer.

Reply:
nice I get what you saying and you're understanding is correct, I also found the code fragile while working with it, will do soon.